INFERENCE.md

```diff
@@ -0,0 +1,202 @@
+# Inference tips
+
+Parler-TTS benefits from a number of optimizations that can make the model up to 4 times faster. Add to this the ability to stream audio as it's being generated, and you can achieve time-to-first-audio in under 500ms on a modern GPU.
+
+## 📖 Quick Index
+* [Efficient Attention Implementation](#efficient-attention-implementations)
+* [Compilation](#compilation)
+* [Streaming](#streaming)
+* [Batch generation](#batch-generation)
+
+## Efficient Attention implementations
+
+Parler-TTS supports [SDPA](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html) and [Flash Attention 2](https://github.com/Dao-AILab/flash-attention).  
+
+SDPA is used by default and speeds up generation time by up to 1.4 times compared with eager attention.
+
+To switch between attention implementations, simply specify `attn_implementation=attn_implementation` when loading the checkpoints:
+
+```py
+from parler_tts import ParlerTTSForConditionalGeneration
+
+torch_device = "cuda:0"
+torch_dtype = torch.bfloat16
+model_name = "parler-tts/parler-tts-mini-v1"
+
+attn_implementation = "eager" # "sdpa" or "flash_attention_2"
+
+model = ParlerTTSForConditionalGeneration.from_pretrained(
+    model_name,
+    attn_implementation=attn_implementation
+).to(torch_device, dtype=torch_dtype)
+```
+
+## Compilation
+
+[Compiling](https://pytorch.org/docs/stable/generated/torch.compile.html) the forward method of Parler can speed up generation time by up to 4.5x.
+
+As an indication, `mode=default` brings a speed-up of 1.4 times compared to no compilation, while a more advanced mode brings much faster generation, at the cost of a longer compilation time and the need to generate twice to see the benefits of compilation.
+
+```py
+import torch
+from parler_tts import ParlerTTSForConditionalGeneration
+from transformers import AutoTokenizer
+
+torch_device = "cuda:0"
+torch_dtype = torch.bfloat16
+model_name = "parler-tts/parler-tts-mini-v1"
+
+# need to set padding max length
+max_length = 50
+
+# load model and tokenizer
+tokenizer = AutoTokenizer.from_pretrained(model_name) 
+model = ParlerTTSForConditionalGeneration.from_pretrained(
+    model_name,
+    attn_implementation="eager"
```

Comment:
IIRC `attn_implementation="sdpa"` with the C++ attn kernel was fastest? Perhaps @eustlb can confirm here!

Reply:
I confirm that the C++ sdpa implementation makes compiled forward with `attn_implementation="sdpa"` fastest. Yet for the mini architecture `attn_implementation="sdpa"`  with compile mode `"default"` and dtype  `bfloat16` is still the fastest.

Reply:
Correction: 
It makes the forward faster only for modes that capture CUDA graphs, i.e `"reduce-overhead"` and `"max-autotune"` but not for "default" where is actually halved generation speed.